### ISSUE TITLE
`unwrap` will return different value when calling `value_and_grads`

### ISSUE THREAD

**Issue by Doraehhh** [google/jax#11377]
`unwrap` will return different value when calling `value_and_grads`. 

```py
import jax
def fn(inp):
    axis = -2
    discont = 2
    period = 1
    return jax.numpy.unwrap(inp, axis=axis, discont=discont, period=period).sum()

mykey = jax.random.PRNGKey(44093535)
inp = jax.random.uniform(mykey, [2, 3, 4], jax.numpy.bfloat16, minval=-128, maxval=127)

res1 = fn(inp)
res2, _ = jax.value_and_grad(fn, (0))(inp)
print(res1)
print(res2)
# -17.0
# -20.0
```

The reason lies in `mod` with `True` as input, which will give differnt value with or without `jit`

```py
def f(dd, period):
    interval = period / 2
    ddmod = jax.numpy.mod(dd + interval, period) - interval
    return ddmod

mykey = jax.random.PRNGKey(44093535)
inp = jax.random.uniform(mykey, [2, 3, 4], jax.numpy.bfloat16, minval=-128, maxval=127)
dd = jax.numpy.diff(inp, axis=-2)
print(f(dd, True))
print(jax.jit(f)(dd, True))
```

```
[[[-0.5 -0.5 0 -0.5]
  [-0.5 -0.5 0 0]]

 [[0 -0.5 0 -0.5]
  [-0.5 0 -0.5 -0.5]]]

[[[-0.5 -0.5  0.   0. ]
  [ 0.   0.   0.   0. ]]

 [[ 0.   0.   0.  -0.5]
  [ 0.   0.  -0.5 -0.5]]]
```

It would be better to convert the `period` argument into `float` at first in `unwrap` to avoid such cases.

**Comment by jakevdp** [comment]
Thanks for the report - it looks like this is not related to `mod`, but rather related to how boolean scalar weak types are determined, which changes type promotion behavior between JIT and non-JIT codepaths:
```python
import jax
import jax.numpy as jnp

def f(x):
  x = jnp.asarray(x)
  print(x.dtype, x.aval.weak_type)

f(True)
# bool False

jax.jit(f)(True)
# bool True
```

**Comment by jakevdp** [comment]
I'm taking a closer look today, and it seems like the examples you show above work correctly in the most recent version of JAX (v0.3.14) – I suspect the changes i #10814 remedied the issue. Can you confirm?

The example I gave above is something I think should be fixed, but it turns out to not be important for this issue in light of the deeper fix in #10814 I think.

**Comment by Doraehhh** [comment]
Yes, it has been fixed in v0.3.14
